### PR TITLE
Process web worker (part 1)

### PR DESCRIPTION
**Pull request by LavrovArtem**
Simple test server:

```js
const http = require('http');

http
    .createServer((req, res) => {
        if (req.url === '/') {
            res.writeHead(200, { 'content-type': 'text/html' });
            res.end(`
                <script>
                    var worker = new Worker('/path/worker.js');
                </script>
            `);
        }
        else if (req.url === '/path/worker.js') {
            res.writeHead(200, { 'content-type': 'application/javascript' });
            res.end(`
                var xhr = new XMLHttpRequest();

                xhr.open('post', '/path/xhr');
                xhr.send();
            `);
        }
        else if (req.url === '/path/xhr') {
            res.writeHead(200, { 'content-type': 'text/plain' });
            res.end('ok');
        }
        else {
            res.writeHead(404);
            res.end();
        }
    })
    .listen(2020);

```

### CODE REVIEW

File: Gulpfile.js [hunk 1]
```diff
@@ -153,7 +154,7 @@ gulp.step('client-scripts-bundle', () => {
         // HACK: babel-plugin-transform-es2015-modules-commonjs forces
         // 'use strict' insertion. We need to remove it manually because
         // of https://github.com/DevExpress/testcafe/issues/258
-        return { code: transformed.code.replace(/^('|")use strict('|");?/, '') };
+        return { code: transformed.code.replace(/^(['"])use strict\1;?/, '') };
```
Comment (miherlosev) [inline Gulpfile.js:157]: Why does this change is necessary?

File: src/client/sandbox/fetch.ts [hunk 2]
```diff
@@ -9,18 +9,15 @@ import SAME_ORIGIN_CHECK_FAILED_STATUS_CODE from '../../request-pipeline/xhr/sam
 import { overrideDescriptor } from '../utils/property-overriding';
 import * as browserUtils from '../utils/browser';
 import { transformHeaderNameToBuiltin, transformHeaderNameToInternal } from '../utils/headers';
+import CookieSandbox from './cookie';
 
-const DEFAULT_REQUEST_CREDENTIALS = nativeMethods.Request ? new nativeMethods.Request(window.location.toString()).credentials : void 0;
+const DEFAULT_REQUEST_CREDENTIALS = nativeMethods.Request ? new nativeMethods.Request(location.toString()).credentials : void 0;
```
Comment (miherlosev) [inline src/client/sandbox/fetch.ts:14]: I propose to create the `getGlobalContextInfo` function and use it in all necessary places.

```js
export default getGlobalContextInfo (): GlobalContextInfo {
    const isInWorker = typeof window === 'undefined';

   return {
         isInWorker,
         global: isInWorker ? self : window
   };
}
```

File: src/client/sandbox/native-methods.ts [hunk 3]
```diff
@@ -688,144 +697,281 @@ class NativeMethods {
...
-        const titleElementTextDescriptor     = win.Object.getOwnPropertyDescriptor(win.HTMLTitleElement.prototype, 'text');
-        const iframeSandboxDescriptor        = win.Object.getOwnPropertyDescriptor(win.HTMLIFrameElement.prototype, 'sandbox');
-        const windowOriginDescriptor         = win.Object.getOwnPropertyDescriptor(win, 'origin');
-
-        if (windowOriginDescriptor) {
-            this.windowOriginGetter = windowOriginDescriptor.get;
-            this.windowOriginSetter = windowOriginDescriptor.set;
-        }
+        // NOTE: this section relates to getting properties from DOM classes
+        if (win.Node) {
```
Comment (miherlosev) [inline src/client/sandbox/native-methods.ts:705]: Let's combine all getters and setters into two separate functions:

```js
this._refreshGetters();
this._refreshSetters();
```

File: src/processing/script/header.ts [hunk 4]
```diff
@@ -20,7 +21,7 @@ const HEADER: string = `
     if (typeof window !== 'undefined' && window){
         window['${INTERNAL_PROPS.processDomMethodName}'] && window['${INTERNAL_PROPS.processDomMethodName}']();
 
-        if (window.${INSTRUCTION.getProperty} && typeof ${INSTRUCTION.getProperty} === 'undefined') {
+        if (window.${INSTRUCTION.getProperty} && typeof ${INSTRUCTION.getProperty} === 'undefined')
```
Comment (miherlosev) [inline src/processing/script/header.ts:24]: For which cases does it necessary?

File: src/processing/script/header.ts [hunk 5]
```diff
@@ -33,20 +34,23 @@ const HEADER: string = `
...
+                ${INSTRUCTION.callMethod} = function(o,p,a){return o[p].apply(o,a)},
+                ${INSTRUCTION.getEval} = function(e){return e},
+                ${INSTRUCTION.processScript} = function(s){return s},
+                ${INSTRUCTION.processHtml} = function(h){return h},
+                ${INSTRUCTION.getPostMessage} = function(w,p){return arguments.length===1?w.postMessage:p},
+                ${INSTRUCTION.getProxyUrl} = function(u,d){return u},
+                ${INSTRUCTION.restArray} = function(a,i){return Array.prototype.slice.call(a, i)},
+                ${INSTRUCTION.restObject} = function(o,p){var k=Object.keys(o),n={};for(var i=0;i<k.length;++i)if(p.indexOf(k[i])<0)n[k[i]]=o[k[i]];return n};
+        
+        if (typeof importScripts !== "undefined" && /\\[native code]/g.test(importScripts.toString()))
```
Comment (miherlosev) [inline src/processing/script/header.ts:52]: Can we simplify template and put this code to the `worker-hammerhead.js`?

Comment (LavrovArtem) [reply src/processing/script/header.ts:52]: https://github.com/DevExpress/testcafe-hammerhead/issues/2404

File: src/processing/script/header.ts (hunk 4 above)
Comment (LavrovArtem) [reply src/processing/script/header.ts:24]: https://github.com/DevExpress/testcafe-hammerhead/issues/2404

File: Gulpfile.js (hunk 1 above)
Comment (miherlosev) [inline Gulpfile.js:157]: This function is called for each client script file. So, it's better to mark it as constant at the start of the `Gulpfile.js`.

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Pull request merged**